src/Nexus.ts

```diff
@@ -246,6 +280,16 @@ class Nexus {
     }
   }
 
+  private setOAuthCredentials(credentials: types.IOAuthCredentials): void {
+    this.mBaseData.headers['Authorization'] = `Bearer ${credentials.token}`;
+    this.mBaseData.cookies['jwt_fingerprint'] = credentials.fingerprint;
+    this.oAuthCredentials = credentials;
+  }
+  
+  public setOnNewOAuthCredentialsHandler(handler: types.OnNewOauthCredentialsHandler) {
+    this.onNewOAuthCredentialsHandler = handler;
```

Comment:
As described above, I would probably maintain a list of event handlers, not force it to be a single one.
Also, again, this is purely about coding style, but I usually just call the function to register an event-handler "on<Event>" so this function could be named "onOAuthCredentialsUpdated" for example.

EDIT: oh wait, I saw further down that "setOnNewOAuthCredentialsHandler" is used internally, is this function even supposed to be public?

Reply:
The method is public. It's for hooking into and receiving new credentials (they can change when the token gets automatically refreshed). This is so a user of this class can persist the new credentials in whatever persistence mechanism they are using.

If I'm making the property containing the handlers an array...should the public setter be an array? Meaning users would have to set all the handlers at once, and if they set it again they lose previous handlers?

Or, should I add the ability to add and remove individual handlers? If so, how would the handlers be referenced?

```
public addOnNewOauthCredentialsHandler(handler: types.INewOauthCredentialsHandler, id: string): void
public removeOnNewOAuthCredentialsHandler(id: string): void
```
???

Reply:
I think you'd _need_ to have a way to add and remove handlers since the class internally uses these callbacks too. How else would you prevent the user-code assignments to the handler-array from overriding the internal one?

You may even go one step further and mirror the api of the nodejs EventEmitter (https://nodejs.org/api/events.html#events_class_eventemitter) and call the functions addListener( / removeListener( with the first parameter specifying which event you're registering for.
That way the api is a bit more familiar to users _and_ we can use the same interface for other events in the future (things like: addListener('request-quota-changed', ...)

Reply:
There is no internal handler.

```
if (this.onNewOAuthCredentialsHandler !== undefined) {
  this.onNewOAuthCredentialsHandler(newOAuthCredentials);
}
```
This just checks whether the handler has been set. If it has, it calls it.

The confusion could have arisen because of the naming, because I used the `on...` prefix for the property?

So, an array is not really _needed_ here at all. But if you still think it's a good idea, why don't I just go the 'full hog' and use `EventEmitter`? (instead of rolling my own solution).

Reply:
Yes, regarding your point about adding other events in future I think I'll integrate a full event system that's standardised so users will be familiar with it. Unless you have objections.

Reply:
Ohh, you're right, I misread the code then.
If you're comfortable turning it into a proper EventEmitter that would probably be a great way to make the api familiar, yes.